sdk/attestation/azure-security-attestation/src/private/attestation_deserializers_private.cpp

```diff
@@ -0,0 +1,297 @@
+        result.PolicyHash,
+        parsedJson,
+        "x-ms-policy-hash",
+        Azure::Core::_internal::Base64Url::Base64UrlDecode);
+
+    JsonOptional::SetIfExists(result.SgxIsDebuggable, parsedJson, "x-ms-sgx-is-debuggable");
+    JsonOptional::SetIfExists(result.SgxProductId, parsedJson, "x-ms-sgx-product-id");
+    JsonOptional::SetIfExists<std::string, std::vector<uint8_t>>(
+        result.SgxMrEnclave, parsedJson, "x-ms-sgx-mrenclave", JsonHelpers::HexStringToBinary);
+    JsonOptional::SetIfExists<std::string, std::vector<uint8_t>>(
+        result.SgxMrSigner, parsedJson, "x-ms-sgx-mrsigner", JsonHelpers::HexStringToBinary);
+    JsonOptional::SetIfExists<std::string, std::vector<uint8_t>>(
+        result.SgxEnclaveHeldData,
+        parsedJson,
+        "x-ms-sgx-ehd",
+        Azure::Core::_internal::Base64Url::Base64UrlDecode);
+    JsonHelpers::SetIfExistsJson(result.SgxCollateral, parsedJson, "x-ms-sgx-collateral");
+    return result;
+  }
+
+  AttestationTokenHeader AttestationTokenHeaderSerializer::Deserialize(
+      Azure::Core::Json::_internal::json const& jsonHeader)
+  {
+    AttestationTokenHeader returnValue;
+
+    JsonOptional::SetIfExists(returnValue.Algorithm, jsonHeader, "alg");
+    JsonOptional::SetIfExists(returnValue.KeyURL, jsonHeader, "jku");
+
+    if (jsonHeader.contains("jwk"))
+    {
+      auto jwk = JsonWebKeySerializer::Deserialize(jsonHeader["jwk"]);
+      returnValue.Key = AttestationSignerInternal(jwk);
+    }
+    JsonOptional::SetIfExists(returnValue.KeyId, jsonHeader, "kid");
+    JsonOptional::SetIfExists(returnValue.X509Url, jsonHeader, "x5u");
+    JsonOptional::SetIfExists(returnValue.X509CertificateChain, jsonHeader, "x5c");
+    JsonOptional::SetIfExists(returnValue.CertificateSha256Thumbprint, jsonHeader, "x5t#S256");
+    JsonOptional::SetIfExists(returnValue.CertificateThumbprint, jsonHeader, "x5t");
+    JsonOptional::SetIfExists(returnValue.Type, jsonHeader, "typ");
+    JsonOptional::SetIfExists(returnValue.ContentType, jsonHeader, "cty");
+    JsonOptional::SetIfExists(returnValue.Critical, jsonHeader, "crit");
+    return returnValue;
+  }
+  std::string AttestationTokenHeaderSerializer::Serialize(AttestationTokenHeader const& tokenHeader)
+  {
+    Azure::Core::Json::_internal::json serializedHeader;
+    JsonOptional::SetFromNullable(tokenHeader.Algorithm, serializedHeader, "alg");
+    JsonOptional::SetFromNullable(tokenHeader.KeyURL, serializedHeader, "jku");
+    JsonOptional::SetFromNullable<AttestationSigner, std::string>(
+        tokenHeader.Key, serializedHeader, "jwk", [](AttestationSigner const& signer) {
+          return AttestationSignerInternal::SerializeToJson(signer);
+        });
+    JsonOptional::SetFromNullable(tokenHeader.ContentType, serializedHeader, "cty");
+    JsonOptional::SetFromNullable(tokenHeader.Critical, serializedHeader, "crit");
+    JsonOptional::SetFromNullable(tokenHeader.KeyId, serializedHeader, "kid");
```

Comment:
nit.
We might want to move all those hardcoded strings to a constants header.  We could re-use many.

Reply:
@ahsonkhan What will autorest be doing for the same constants? Will it be using manifest constants or creating a separate manifest constant header? Since this is a functional replacement for autorest generated code (and will be removed once the autorest serializers are created), I want to follow the pattern autorest is going to use.